console-subscriber/src/aggregator/mod.rs

```diff
@@ -368,19 +387,28 @@ impl Aggregator {
         let stats = &self.task_stats;
         // Assuming there are much fewer task details subscribers than there are
         // stats updates, iterate over `details_watchers` and compact the map.
-        self.details_watchers.retain_and_shrink(|id, watchers| {
-            if let Some(task_stats) = stats.get(id) {
-                let details = proto::tasks::TaskDetails {
-                    task_id: Some(id.clone().into()),
-                    now: Some(self.base_time.to_timestamp(Instant::now())),
-                    poll_times_histogram: task_stats.serialize_histogram(),
-                };
-                watchers.retain(|watch| watch.update(&details));
-                !watchers.is_empty()
-            } else {
-                false
-            }
-        });
+        self.details_watchers
+            .retain_and_shrink(|id, updates_and_watchers| {
+                if let Some(task_stats) = stats.get(id) {
+                    let details = proto::tasks::TaskDetails {
+                        task_id: Some(id.clone().into()),
+                        now: Some(self.base_time.to_timestamp(Instant::now())),
+                        poll_times_histogram: task_stats.serialize_histogram(),
+                        causality: vec![],
+                    };
+                    updates_and_watchers.retain(|(updates, watch)| {
+                        let updates = crate::consequences::updates_into_proto(updates);
+                        let details = proto::tasks::TaskDetails {
+                            causality: updates,
+                            ..details.clone()
+                        };
```

Comment:
maybe make `details` mutable so we can just update it and avoid the clone? or flip the order of the code so that we have the updates variable available before constructing `details`